mlir/lib/Catalyst/Transforms/QnodeToAsyncPatterns.cpp

```diff
@@ -0,0 +1,101 @@
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "scatter"
+
+#include <algorithm>
+#include <vector>
+
+#include "mlir/Dialect/Async/IR/Async.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+
+namespace catalyst {
+
+struct CallOpToAsyncOPRewritePattern : public mlir::OpRewritePattern<func::CallOp> {
+    using mlir::OpRewritePattern<func::CallOp>::OpRewritePattern;
+
+    mlir::LogicalResult matchAndRewrite(func::CallOp op,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        SymbolRefAttr symbol = dyn_cast_if_present<SymbolRefAttr>(op.getCallableForCallee());
+        func::FuncOp func = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, symbol);
+        if (!func) {
+            // Nothing to change.
+            return failure();
+        }
+
+        // Check for Call ops that have QNode func ops
+        if (!func->hasAttrOfType<UnitAttr>("qnode")) {
+            // Nothing to change. (For functions which are not qnodes).
+            return failure();
+        }
+
+        if (op->hasAttrOfType<UnitAttr>("transformed")) {
+            // Nothing to change. (For functions which are not qnodes).
+            return failure();
+        }
+
+        TypeRange retTy = op.getResultTypes();
+        SmallVector<Value> dependencies; /* = empty */
+        SmallVector<Value> operands;     /* = empty */
+        auto noopExec = [&](OpBuilder &executeBuilder, Location executeLoc,
+                            ValueRange executeArgs) {};
+
+        rewriter.updateRootInPlace(op, [&] { op->setAttr("transformed", rewriter.getUnitAttr()); });
+        IRMapping map;
+        auto executeOp =
+            rewriter.create<async::ExecuteOp>(op.getLoc(), retTy, dependencies, operands, noopExec);
+        {
+            PatternRewriter::InsertionGuard insertGuard(rewriter);
+            rewriter.setInsertionPoint(executeOp.getBody(), executeOp.getBody()->end());
+            Operation *cloneOp = op->clone(map);
+            rewriter.insert(cloneOp);
+            rewriter.create<async::YieldOp>(op.getLoc(), cloneOp->getResults());
+        }
+
+        auto asyncValues = executeOp.getResults();
+
+        // We should move this as much as we can...
+        rewriter.create<async::AwaitOp>(op.getLoc(), asyncValues.front());
+        rewriter.create<async::RuntimeDropRefOp>(op.getLoc(), asyncValues.front(),
```

Comment:
Why do you treat the firs one differently?

Reply:
[From MLIR's documentation](https://mlir.llvm.org/docs/Dialects/AsyncDialect/#asyncexecute-asyncexecuteop)

```llvm
%token, %results =
  async.execute [%dependency](%value as %unwrapped: !async.value<f32>)
             -> !async.value<!some.type>
  {
    %0 = "compute0"(%unwrapped): (f32) -> !some.type
    async.yield %0 : !some.type
  }
```

`async.execute` preppends an `async.token` to the results of `async.execute`.